### ISSUE TITLE
QSO macro models stall in scatter when optical depth is high

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#82]
**Brief Description:** code spends a lot of time choosing directions for line photons in optically thick regions

**How To Reproduce:** Fiducial model, data set h10_standard73 or similar, macro atom line transfer.

**Discussion:**In Issue #43 I highlighted the inability of the macro atom models to work with QSO models in ionization mode 7. I think there are 2 separate problems here, one of which is detailed there but one of which has nothing to do with ionization. When running QSO models, the code will stall after a while in scatter. The time usage after a while looks something like the following (99.7% of time spent in scatter!):

![screenshot 2014-06-11 13 35 27](https://cloud.githubusercontent.com/assets/3007249/3248114/077b8fa4-f18b-11e3-8252-53921344306f.png)

I've traced this to a lot of time spent in a series of loops where one decides the new direction of a line photon in the 'anisotropic scattering model', i.e. where we pick the line photon direction based on the chance of it escaping the sobolev region on each line of sight. The code that does this looks like this:

``` C
      ztest = 1.0;
      z = 0.0;
      *nnscat = *nnscat - 1;
      while (ztest > z)
    {
      *nnscat = *nnscat + 1;
      randvec (z_prime, 1.0);   /* Get a new direction for the photon (isotropic */
      stuff_v (z_prime, p->lmn);
      ztest = (rand () + 0.5) / MAXRAND;
      dvds = dvwind_ds (p);
      ishell = p->grid;
      tau = sobolev (&wmain[ishell], p, -1.0, lin_ptr[p->nres], dvds);
      if (tau < 1.0e-5)
        z = 1.0;
      else
        z = (1. - exp (-tau)) / tau;    /* probability to see if it escapes in that direction */
    }
```

What I think is happening is that in very dense regions of the wind one collisionally excites a macro atom. The chance of there being a radiative de-excitation is pretty rare, becuase you have low p_escape so the radiative downwards rates are small, and other cooling processes may dominate, but occasionally it happens. 

When this happens each value of z in the above loop is very small because the line is optically thick in all directions, and so it takes ages to get a direction. Do you agree @ssim ? I'm also not sure why nnscat gets incremented in this loo. I

**Possible Solutions:** I feel there must be a better way to do this- maybe build up a CDF of the directions, go through it til the random direction is reach then interpolate so that we don't have resolution limited angles?

I also chucked in a quick fudge to see if changing this speeds things up- everytime the overall P-escape was low I just said choose the direction isotropically, because we don't really care about those photons, and this speeds things up a lot, and may even give

**Comment by jhmatthews** [comment]
I've checked, and this code isn't actually duplicated anywhere. According to some commenting it was moved from trans_phot to here, and is only called either in macro atom mode or when we use a specific line transfer mode, which hasn't been used for ages. It is my understanding therefore that there are three ways of dealing with sobolev optical depths in the code
- for non macro-atom mode, resonance scattering is done by full MC, so the emergent direction is taken care of by actually doing the scatters. 
- for macro atoms _and_ simple ions, we use the loop above to choose the anisotropy such that the photon direction is chosen according to the local sobolev optical depth.
- for thermally produced lines in non matom mode, we adjust the line radiation by p_escape, and seemingly 
- there's also anisotropic mode 1, which I don't fully understand, and would like to discuss next time we talk.
- I also am not sure I understand why in macro atoms we choose the direction based on sobolev tau, but don't do this in the normal mode.

**Proposed Fix**

I've been thinking about this, and I think this is the best way to do this. 
- Really you want your random number to be normalized to the maximum probability of escaping. That is, instead of normalizing to 1.0, you normalize to whatever the highest value of (1. - exp (-tau)) / tau is along each direction.
- One could just choose a load of random directions and record the maximum
- but being as we know tau is proportional to dv/ds I think I'm right in saying that the maximum escape probability must be somewhere along either the vector of the poloidal velocity or the vector of the rotational velocity. 
- so we can just evaluate (1. - exp (-tau)) / taualong each of those, pick the largest and normalise to that value.

Sound sensible, @kslong , @ssim , @lazygun37 ?

**Comment by lazygun37** [comment]
Hi James,

I don't think your 3rd and 4th point under "proposed fix" are correct -- it is not generically true that the max gradient has to be along poloidal or rotational directions. For example, consider a wind with no rotation at all and a slowly accelerating KWD style velocity law. Suppose the poloidal acceleration near the base of the wind is really _extremely_ slow, i.e. very little poloidal gradient. But, in the KWD law, the starting velocity scales as thermal speed, which scales as T^0.5, which does scale radially across the disk. So it is conceivable that the gradient in such a case could be larger _across_ poloidal streamlines than along them. Similarly, once you get far enough away that you reach poloidal terminal velocity, the max gradient is almost guaranteed to lie _across_ the poloidal direction, i.e. cylindrically outward. If you add in rotational velocities, things just get more complicated, I think, and it's not always obvious where the max dv/ds is going to lie.

Cheers,

C

**Comment by jhmatthews** [comment]
Yes of course, I was being a bit silly and only thinking about a streamline in isolation and forgetting about the velocity gradient between streamlines both in x and z directions.

**Comment by jhmatthews** [comment]
It turns out we store the maximum dv/ds in a cell. So I think we can just do the following 

``` C
      tau_norm = sobolev (&wmain[n], p, -1.0, lin_ptr[p->nres], wmain[n].dvds_max);
      p_norm = (1. - exp (-tau_norm)) / tau_norm;


      while (ztest > z)
    {
      *nnscat = *nnscat + 1;
      randvec (z_prime, 1.0);   /* Get a new direction for the photon (isotropic */
      stuff_v (z_prime, p->lmn);
      ztest = (rand () + 0.5) / MAXRAND * p_norm;
      dvds = dvwind_ds (p);
      ishell = p->grid;
      tau = sobolev (&wmain[ishell], p, -1.0, lin_ptr[p->nres], dvds);
      if (tau < 1.0e-5)
        z = 1.0;
      else
        z = (1. - exp (-tau)) / tau;    /* probability to see if it escapes in that direction */
    }
```

so that ztest is normalised by the escape probability along the maximum dvds. This dvds_max is worked out when we calculate the average dvds, by choosing 10000 directions and storing the maximum, so there's a resolution issue there- there could be a direction which isn't sampled which has a higher value. 

However, is it really going to be an issue? It's not obvious to me but I doubt it. We can also make sure that we only do this when p_escape is low, so otherwise it just uses 1.0 as the normalisation.

**Comment by kslong** [comment]
I am confused by this.  In particular, I am quite sure we do choose the direction based on sobolev Tau and dv/ds in the ``normal mode’’.  You seem to be indicating our normal mode has isotropic scattering.  I thought this was all pretty well documented in the code.  I agree though we should go through all of the scattering modes and make sure we understand how they are implemented.

Knox

**Comment by jhmatthews** [comment]
I was also confused by this Knox. In python.c, when line transfer mode is 3 (which is 'normal') then we go into the following statement

``` C
else
    {
      geo.scatter_mode = 0; // isotropic
      geo.rt_mode = 1;      // Not macro atom (SS)
    }
```

This option means that one then calls randvec to generate a new direction in scatter. This is also done when one generates wind photons.

**Comment by kslong** [comment]
Well, if that is the case, we should stop using this mode.  Long and Knigge did not use isotropic scattering.  Knox

**Comment by jhmatthews** [comment]
This is a little worrying. It seems to me that for a while now we have been using line transfer mode 3, which sets scattering to isotropic. This is definitely what was used in the benchmark QSO model (I've checked which routines are entered), and I suspect also in the Proga model. It's hardly surprising this happened, as most pf files don't even list 4,5 as options, and 3 does not mention isotropic or anisotropic.

I've rerun a benchmark model with line transfer mode set to 4, which uses anisotropic scattering mode 1 (see below). Some small differences, but we are mostly OK.

![comp_run110e_run110e_mode4](https://cloud.githubusercontent.com/assets/3007249/3303739/9d4fb6c0-f642-11e3-99f8-47dac93feada.png)

I suggest @Higginbottom runs the Proga model ASAP.

**Scattering modes**

Isotropic scattering, scatter mode 0, is currently used by the standard line_mode=3  we'd been using so far. 

The line_mode=4 which I assume is what was used in LK02 uses another anisotropic scattering mode 1. I haven't quite figured out how this works yet but seems to generate a pdf of the function vrandwind and then choose a direction from that.

In the above bug, I was talking about anisotropic scattering mode 2 which uses the above rejection method to choose the direction, and is used in macro atom mode, line_mode=7.

Hope that made some kind of sense. Plenty to talk about...

**Comment by kslong** [comment]
I don’t think there is really any reason to worry much at present.  I don’t personally think Nick needs to do anything at this point.   It does suggest to me  that we need a working manual which explains what the various working options mean, and that we should decide what the defaults in the program should be and why.

Knox

**Comment by jhmatthews** [comment]
Ok Knox. I've created a page on the github wiki [here](https://github.com/agnwinds/python/wiki/Line-Transfer-and-Scattering) to describe the line transfer and scattering modes. I'm still unsure how scattering mode 1 works, which I think is the one used in LK02. Something to discuss tomorrow, particular relating to what we want the defaults to be.

I'm also fairly happy I have a good fix for the original issue and have been doing a little work on how dvds tends to vary in a cell, but we can double check tomorrow before I merge the relevant changes.

**Comment by jhmatthews** [comment]
The fix for the original issue is now implemented, I am running some tests and have created an extra line transfer mode which is documented on the wiki. The changes are documented in #83

**Comment by jhmatthews** [comment]
Fixed by #83. Closed.
